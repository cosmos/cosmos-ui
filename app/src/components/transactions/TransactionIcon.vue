<template>
  <div class="tx__icon">
    <img :src="icon" :alt="transactionType" />
  </div>
</template>

<script>
import config from "src/../config"
export default {
  name: `transaction-icon`,
  props: {
    transactionType: {
      type: String,
      required: true,
    },
  },
  data: () => ({
    isExtension: config.isExtension,
  }),
  computed: {
    icon() {
      return this.isExtension
        ? `/icons/activity/${this.transactionType}.svg`
        : `/img/icons/activity/${this.transactionType}.svg`
    },
  },
}
</script>

<style>
.tx__icon {
  padding: 1rem;
}

.tx__icon img {
  max-height: 100%;
  max-width: 40px;
  display: block;
}

@media screen and (min-width: 1023px) {
  .tx__icon img {
    max-width: 48px;
  }
}
</style>
