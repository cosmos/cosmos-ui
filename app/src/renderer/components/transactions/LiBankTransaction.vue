<template>
  <li-transaction
    :color="`#ED553B`"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="sent">
      <div slot="caption">
        Sent&nbsp;
        <b>{{ full(atoms(coins.amount)) }}</b>
        <span>&nbsp;{{ coins.denom.toUpperCase() }}</span>
      </div>
      <span slot="details">
        <template v-if="sentSelf">
          To yourself!
        </template><template v-else>
          To {{ receiver }}
        </template>
      </span>
    </template><template v-else>
      <div slot="caption">

        Received&nbsp;
        <b>{{ full(atoms(coins.amount)) }}</b>
        <span>&nbsp;{{ coins.denom.toUpperCase() }}</span>
      </div>
      <span slot="details">From {{ sender }}</span>
    </template>
  </li-transaction>
</template>

<script>
import LiTransaction from "./LiTransaction"
import { atoms, full } from "../../scripts/num.js"

export default {
  name: `li-bank-transaction`,
  components: { LiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    address: {
      type: String,
      default: null
    }
  },
  data: () => ({
    atoms,
    full
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    },
    // TODO: sum relevant inputs/outputs
    sentSelf() {
      return this.tx.from_address === this.tx.to_address
    },
    sent() {
      return this.tx.from_address === this.address
    },
    sender() {
      return this.tx.from_address
    },
    coins() {
      return this.tx.amount[0]
    },
    receiver() {
      return this.tx.to_address
    }
  }
}
</script>
