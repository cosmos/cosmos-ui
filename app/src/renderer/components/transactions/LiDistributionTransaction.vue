<template>
  <li-transaction
    :color="`#F2B134`"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="txType === `cosmos-sdk/MsgWithdrawDelegationReward`">
      <div slot="caption">
        Withdraw rewards
      </div>
      <div slot="details">
        From&nbsp;<router-link :to="url + '/' + tx.validator_addr">
          {{
            moniker(tx.validator_addr)
          }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgSetWithdrawAddress`">
      <div slot="caption">
        Update withdraw address
      </div>
      <div slot="details">
        To {{ tx.withdraw_address }}
      </div>
    </template>
    <template
      v-else-if="txType === `cosmos-sdk/MsgWithdrawValidatorCommission`"
    >
      <div slot="caption">
        Withdraw validator commission
      </div>
      <div slot="details">
        From&nbsp;<router-link :to="url + '/' + tx.validator_address">
          {{
            moniker(tx.validator_address)
          }}
        </router-link>
      </div>
    </template>
  </li-transaction>
</template>

<script>
import LiTransaction from "./LiTransaction"
import { pretty, atoms } from "../../scripts/num.js"

export default {
  name: `li-distribution-transaction`,
  components: { LiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    url: {
      type: String,
      required: true
    },
    bondingDenom: {
      type: String,
      required: true
    },
    txType: {
      type: String,
      required: true
    },
    validators: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    atoms,
    pretty
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    }
  },
  methods: {
    moniker(validatorAddr) {
      const validator = this.validators.find(
        c => c.operator_address === validatorAddr
      )
      return validator ? validator.description.moniker : validatorAddr
    }
  }
}
</script>
