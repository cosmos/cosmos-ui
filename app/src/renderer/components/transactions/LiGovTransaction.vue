<template>
  <li-transaction
    :color="`#15CFCC`"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="txType === `cosmos-sdk/MsgSubmitProposal`">
      <div slot="caption">
        Submit {{ tx.proposal_type.toLowerCase() }} proposal&nbsp;
        <b>{{ atoms(tx.initial_deposit[0].amount) }}</b>
        <span>&nbsp;{{ tx.initial_deposit[0].denom }}s</span>
      </div>
      <div slot="details">
        Title:&nbsp;<i>{{ tx.title }}</i>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgDeposit`">
      <div slot="caption">
        Deposit&nbsp;
        <template>
          <b>{{ pretty(atoms(tx.amount[0].amount)) }}</b>
          <span>&nbsp;{{ tx.amount[0].denom }}s</span>
        </template>
      </div>
      <div slot="details">
        On&nbsp;
        <router-link :to="`${url}/${tx.proposal_id}`">
          Proposal &#35;{{ tx.proposal_id }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgVote`">
      <div slot="caption">
        Vote&nbsp;<b>{{ tx.option }}</b>
      </div>
      <div slot="details">
        On&nbsp;
        <router-link :to="`${url}/${tx.proposal_id}`">
          Proposal &#35;{{ tx.proposal_id }}
        </router-link>
      </div>
    </template>
  </li-transaction>
</template>

<script>
import LiTransaction from "./LiTransaction"
import { pretty, atoms } from "../../scripts/num.js"

export default {
  name: `li-gov-transaction`,
  components: { LiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    url: {
      type: String,
      required: true
    },
    bondingDenom: {
      type: String,
      required: true
    },
    txType: {
      type: String,
      required: true
    }
  },
  data: () => ({
    pretty,
    atoms
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    }
  }
}
</script>
