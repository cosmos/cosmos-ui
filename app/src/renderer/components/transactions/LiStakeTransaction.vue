<template>
  <li-transaction
    color="#47AB6C"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="txType === `cosmos-sdk/MsgCreateValidator`">
      <div slot="caption">
        Create validator&nbsp;
        <b>{{ full(atoms(tx.value.amount)) }}</b>
        <span>&nbsp;{{ tx.value.denom }}s</span>
      </div>
      <div slot="details">
        Moniker:&nbsp;
        <router-link :to="`${url}/${tx.validator_address}`">
          {{
            moniker(tx.validator_address)
          }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgEditValidator`">
      <div slot="caption">
        Edit validator&nbsp;
        <router-link :to="`${url}/${tx.validator_address}`">
          {{
            moniker(tx.validator_address)
          }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgDelegate`">
      <div slot="caption">
        Delegation&nbsp;
        <b>{{ full(atoms(tx.value.amount)) }}</b>
        <span>&nbsp;{{ tx.value.denom }}s</span>
      </div>
      <div slot="details">
        To&nbsp;
        <router-link :to="`${url}/${tx.validator_address}`">
          {{
            moniker(tx.validator_address)
          }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgBeginRedelegate`">
      <div slot="caption">
        Redelegation&nbsp;
        <b>
          {{
            calculatePrettifiedTokens(tx.validator_src_address, tx.shares_amount)
          }}
        </b>
        <span>&nbsp;{{ bondingDenom }}s</span>
      </div>
      <div slot="details">
        From&nbsp;
        <router-link :to="`${url}/${tx.validator_src_address}`">
          {{
            moniker(tx.validator_src_address)
          }}
        </router-link>
        &nbsp;to&nbsp;
        <router-link :to="`${url}/${tx.validator_dst_address}`">
          {{
            moniker(tx.validator_dst_address)
          }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgUndelegate`">
      <div slot="caption">
        Undelegation&nbsp;
        <b>
          {{
            calculatePrettifiedTokens(tx.validator_address, tx.shares_amount)
          }}
        </b>
        <span>&nbsp;{{ bondingDenom }}s</span>
        <template v-if="timeDiff">
          <span class="tx-unbonding__time-diff">
            &nbsp;{{ timeDiff }}
          </span>
        </template>
      </div>
      <div slot="details">
        From&nbsp;
        <router-link :to="`${url}/${tx.validator_address}`">
          {{
            moniker(tx.validator_address)
          }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgUnjail`">
      <div slot="caption">
        Unjail&nbsp;
        <router-link :to="`${url}/${tx.address}`">
          {{
            moniker(tx.address)
          }}
        </router-link>
      </div>
    </template>
  </li-transaction>
</template>

<script>
import LiTransaction from "./LiTransaction"
import { atoms, full } from "../../scripts/num.js"
import { calculateTokens } from "../../scripts/common.js"
import moment from "moment"

/*
 * undelegation tx need a preprocessing, where shares are translated into transaction.balance: {amount, denom}
 */

export default {
  name: `li-stake-transaction`,
  components: { LiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    validators: {
      type: Array,
      required: true
    },
    url: {
      type: String,
      required: true
    },
    bondingDenom: {
      type: String,
      required: true
    },
    unbondingTime: {
      type: Number,
      default: null
    },
    txType: {
      type: String,
      required: true
    }
  },
  data: () => ({
    atoms,
    full
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    },
    timeDiff() {
      // only show time diff if still waiting to be terminated
      if (this.state !== `locked`) return ``

      return `(liquid ${moment(this.unbondingTime).fromNow()})`
    },
    // unbonding transactions can be in the state 'locked', 'ended'
    // the transaction needs to be enriched from the outside with `unbondingDelegation`
    state() {
      if (!this.unbondingTime) return `ended`
      if (this.unbondingTime - Date.now() <= 0) return `ended`
      return `locked`
    }
  },
  methods: {
    moniker(validatorAddr) {
      const validator = this.validators.find(
        c => c.operator_address === validatorAddr
      )
      return validator ? validator.description.moniker : validatorAddr
    },
    calculatePrettifiedTokens(validatorAddr, shares) {
      const validator = this.validators.find(
        val => val.operator_address === validatorAddr
      )
      return this.atoms(calculateTokens(validator, shares).toNumber())
    }
  }
}
</script>
