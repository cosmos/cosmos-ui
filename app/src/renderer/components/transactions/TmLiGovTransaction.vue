<template>
  <tm-li-transaction
    :color="color"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="propose">
      <div slot="caption">
        Submitted {{ tx.proposal_type.toLowerCase() }} proposal initial
        deposit&nbsp;<b>{{ pretty(tx.initial_deposit[0].amount) }}</b>
        <span>&nbsp;{{ tx.initial_deposit[0].denom }}s</span>
      </div>
      <div slot="details">
        Title:&nbsp;<i>{{ tx.title }}</i>
      </div>
    </template>
    <template v-if="deposit">
      <div slot="caption">
        Deposited&nbsp;
        <template>
          <b>{{ pretty(tx.amount[0].amount) }}</b>
          <span>&nbsp;{{ tx.amount[0].denom }}s</span>
        </template>
      </div>
      <div slot="details">
        On&nbsp;
        <router-link :to="url + '/' + tx.proposal_id">
          Proposal &#35;{{ tx.proposal_id }}
        </router-link>
      </div>
    </template>
  </tm-li-transaction>
</template>

<script>
import TmLiTransaction from "./TmLiTransaction"
import colors from "./transaction-colors.js"
import { pretty } from "../../scripts/num.js"

export default {
  name: `tm-li-gov-transaction`,
  components: { TmLiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    url: {
      type: String,
      required: true
    },
    bondingDenom: {
      type: String,
      required: true
    }
  },
  data: () => ({
    pretty
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    },
    type() {
      return this.transaction.tx.value.msg[0].type
    },
    propose() {
      return this.type === `cosmos-sdk/MsgSubmitProposal`
    },
    deposit() {
      return this.type === `cosmos-sdk/MsgDeposit`
    },
    color() {
      if (this.propose) return colors.gov.propose
      return colors.gov.deposit
    }
  }
}
</script>

<style>
.tm-li-tx {
  display: flex;
  font-size: sm;
}

.tm-li-tx .tx-icon {
  padding: 0 0.5rem;
  background: var(--app-fg);
  display: flex;
  align-items: center;
  justify-content: center;
}

.tm-li-tx .tx-container {
  flex-direction: column;
  flex-wrap: nowrap;
  padding: 0.5rem 0;
  margin: 0.5rem 0;
  display: flex;
  width: 100%;
  min-width: 0;
}

.tm-li-tx .tx-element {
  padding: 0 2rem 0 1.5rem;
  line-height: 1.5rem;
}

.tm-li-tx .tx-coin .value {
  flex: 0 0 100%;
  font-size: var(--sm);
  color: var(--dim);
}

.tm-li-tx .tx-coin .value:before {
  content: "";
  display: inline;
}

.tm-li-tx .tx-coin .key {
  font-weight: 500;
  font-size: var(--m);
}

.tm-li-tx .tx-coin .value,
.tm-li-tx .tx-coin .key {
  line-height: 1.5rem;
}

.tm-li-tx.tm-li-tx-sent .tx-coin .value:before {
  content: "-";
}

.tm-li-tx.tm-li-tx-received .tx-icon {
  background: var(--app-fg);
}

.tm-li-tx.tm-li-tx-received .tx-coin .value {
  color: var(--success);
}

.tm-li-tx.tm-li-tx-received .tx-coin .value:before {
  content: "+";
}

.tm-li-tx:hover {
  cursor: pointer;
}

@media screen and (min-width: 700px) {
  .tm-li-tx {
    font-size: 0.875rem;
  }

  .tm-li-tx .tx-container {
    flex-direction: row;
  }

  .tm-li-tx .tx-container .tx-coins {
    flex: 0 0 9rem;
    padding: 0;
    min-width: 0;
  }

  .tm-li-tx .tx-container .tx-coins .tx-coin {
    padding: 0 1.5rem 0;
  }

  .tm-li-tx .tx-container .tx-coins .tx-coin .key {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
}
</style>
