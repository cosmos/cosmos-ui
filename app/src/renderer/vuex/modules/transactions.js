import fp from "lodash/fp"
import { uniqBy } from "lodash"
import * as Sentry from "@sentry/browser"
import Vue from "vue"
export default ({ node }) => {
  let emptyState = {
    loading: false,
    loaded: false,
    error: null,
    wallet: [], // {height, result: { gas, tags }, tx: { type, value: { fee: { amount: [{denom, amount}], gas}, msg: {type, inputs, outputs}}, signatures} }}
    staking: [],
    governance: []
  }
  let state = JSON.parse(JSON.stringify(emptyState))

  // properties under which txs of different categories are stored
  const txCategories = [`staking`, `wallet`, `governance`]

  let mutations = {
    setWalletTxs(state, txs) {
      state.wallet = txs
    },
    setStakingTxs(state, txs) {
      state.staking = txs
    },
    setGovernanceTxs(state, txs) {
      state.governance = txs
    },
    setHistoryLoading(state, loading) {
      state.loading = loading
    },
    setTransactionTime(state, { blockHeight, blockMetaInfo }) {
      txCategories.forEach(category => {
        state[category].forEach(t => {
          if (t.height === blockHeight && blockMetaInfo) {
            // time seems to be an ISO string, but we are expecting a Number type
            Vue.set(t, `time`, new Date(blockMetaInfo.header.time).getTime())
          }
        })
      })
    }
  }

  let actions = {
    resetSessionData({ rootState }) {
      // clear previous account state
      rootState.transactions = JSON.parse(JSON.stringify(emptyState))
    },
    async reconnected({ state, dispatch, rootState }) {
      if (state.loading && rootState.user.signedIn) {
        await dispatch(`getAllTxs`)
      }
    },
    async getAllTxs({ commit, dispatch, state, rootState }) {
      try {
        state.loading = true

        if (!rootState.connection.connected) return

        const stakingTxs = await dispatch(`getTx`, `staking`)
        commit(`setStakingTxs`, stakingTxs)

        const governanceTxs = await dispatch(`getTx`, `governance`)
        commit(`setGovernanceTxs`, governanceTxs)

        const walletTxs = await dispatch(`getTx`, `wallet`)
        commit(`setWalletTxs`, walletTxs)

        const allTxs = stakingTxs.concat(governanceTxs.concat(walletTxs))
        await dispatch(`enrichTransactions`, {
          transactions: allTxs
        })
        state.error = null
        state.loading = false
        state.loaded = true
      } catch (error) {
        commit(`notifyError`, {
          title: `Error getting transactions`,
          body: error.message
        })
        Sentry.captureException(error)
        state.error = error
      }
    },
    async getTx(
      {
        rootState: {
          user: { address }
        }
      },
      type
    ) {
      let response
      switch (type) {
        case `staking`:
          response = await node.getDelegatorTxs(address)
          break
        case `governance`:
          response = await node.getGovernanceTxs(address)
          break
        case `wallet`:
          response = await node.txs(address)
          break
        default:
          throw new Error(`Unknown transaction type`)
      }
      const transactionsPlusType = response.map(fp.set(`type`, type))
      return response ? uniqBy(transactionsPlusType, `hash`) : []
    },
    async enrichTransactions({ dispatch }, { transactions }) {
      const blockHeights = new Set(
        transactions.map(({ height }) => parseInt(height))
      )
      await Promise.all(
        [...blockHeights].map(blockHeight =>
          dispatch(`queryTransactionTime`, { blockHeight })
        )
      )
    },
    async queryTransactionTime({ commit, dispatch }, { blockHeight }) {
      let blockMetaInfo = await dispatch(`queryBlockInfo`, blockHeight)
      commit(`setTransactionTime`, {
        blockHeight,
        blockMetaInfo
      })
    }
  }

  return {
    state,
    mutations,
    actions
  }
}
