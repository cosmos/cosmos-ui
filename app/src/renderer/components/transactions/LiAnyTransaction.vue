<template>
  <div>
    <div
      v-for="(msg, index) in transaction.tx.value.msg"
      :key="index"
    >
      <li-bank-transaction
        v-if="bankTx(msg.type)"
        :tx="msg.value"
        :bonding-denom="bondingDenom"
        :address="address"
        :fees="fees"
        :time="transaction.time"
        :block="Number(transaction.height)"
      />
      <li-stake-transaction
        v-else-if="stakingTx(msg.type)"
        :tx="msg.value"
        :validators="validators"
        :url="validatorsUrl"
        :unbonding-time="unbondingTime"
        :bonding-denom="bondingDenom"
        :tx-type="msg.type"
        :fees="fees"
        :time="transaction.time"
        :block="Number(transaction.height)"
      />
      <li-gov-transaction
        v-else-if="governanceTx(msg.type)"
        :tx="msg.value"
        :bonding-denom="bondingDenom"
        :url="proposalsUrl"
        :tx-type="msg.type"
        :fees="fees"
        :time="transaction.time"
        :block="Number(transaction.height)"
      />
      <li-distribution-transaction
        v-else-if="distributionTx(msg.type)"
        :tx="msg.value"
        :url="validatorsUrl"
        :bonding-denom="bondingDenom"
        :tx-type="msg.type"
        :validators="validators"
        :fees="fees"
        :time="transaction.time"
        :block="Number(transaction.height)"
      />
      <li-transaction
        v-else
        :time="transaction.time"
        :block="Number(transaction.height)"
        color="grey"
      >
        <span slot="caption">Unknown Transaction Type</span>
      </li-transaction>
    </div>
  </div>
</template>

<script>
import LiBankTransaction from "./LiBankTransaction"
import LiStakeTransaction from "./LiStakeTransaction"
import LiGovTransaction from "./LiGovTransaction"
import LiDistributionTransaction from "./LiDistributionTransaction"
import LiTransaction from "./LiTransaction"

export default {
  name: `li-any-transaction`,
  components: {
    LiBankTransaction,
    LiGovTransaction,
    LiStakeTransaction,
    LiDistributionTransaction,
    LiTransaction
  },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    address: {
      type: String,
      default: null
    },
    validators: {
      type: Array,
      required: true
    },
    validatorsUrl: {
      type: String,
      default: null
    },
    proposalsUrl: {
      type: String,
      default: null
    },
    bondingDenom: {
      type: String,
      required: true
    },
    unbondingTime: {
      type: Number,
      default: null
    }
  },
  computed: {
    fees() {
      return (
        this.transaction.tx.value.fee.amount &&
        this.transaction.tx.value.fee.amount[0]
      )
    }
  },
  methods: {
    bankTx(type) {
      return [`cosmos-sdk/MsgSend`].includes(type)
    },
    stakingTx(type) {
      return [
        `cosmos-sdk/MsgCreateValidator`,
        `cosmos-sdk/MsgEditValidator`,
        `cosmos-sdk/MsgDelegate`,
        `cosmos-sdk/MsgUndelegate`,
        `cosmos-sdk/MsgBeginRedelegate`,
        `cosmos-sdk/MsgUnjail`
      ].includes(type)
    },
    governanceTx(type) {
      return [
        `cosmos-sdk/MsgSubmitProposal`,
        `cosmos-sdk/MsgDeposit`,
        `cosmos-sdk/MsgVote`
      ].includes(type)
    },
    distributionTx(type) {
      return [
        `cosmos-sdk/MsgSetWithdrawAddress`,
        `cosmos-sdk/MsgWithdrawDelegationReward`,
        `cosmos-sdk/MsgWithdrawValidatorCommission`
      ].includes(type)
    }
  }
}
</script>
