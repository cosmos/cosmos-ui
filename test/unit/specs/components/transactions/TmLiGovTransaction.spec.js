import { mount } from "@vue/test-utils"
import TmLiGovTransaction from "transactions/TmLiGovTransaction"
import transactions from "../../store/json/txs"

describe(`TmLiGovTransaction`, () => {
  let wrapper
  const propsData = {
    transaction: transactions[6],
    url: `/proposals`,
    bondingDenom: `stake`
  }

  beforeEach(() => {
    wrapper = mount(TmLiGovTransaction, { propsData, stubs: [`router-link`] })
  })

  describe(`proposals`, () => {
    it(`should show proposals submission transactions`, () => {
      expect(wrapper.vm.propose).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })

  describe(`deposits`, () => {
    it(`should show deposits`, () => {
      wrapper.setProps({
        transaction: transactions[7]
      })
      expect(wrapper.vm.deposit).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})
