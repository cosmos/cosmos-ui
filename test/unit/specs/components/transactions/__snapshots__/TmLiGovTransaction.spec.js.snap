// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmLiGovTransaction deposits should show deposits 1`] = `
<div
  class="tm-li-tx"
>
  <div
    class="tm-li-tx__icon"
  >
    <img
      src="~assets/images/cosmos-logo.png"
      style="border-color: #5effe9;"
    />
  </div>
   
  <div
    class="tm-li-tx__content"
  >
    <div
      class="tm-li-tx__content__left"
    >
      <div
        class="tm-li-tx__content__caption"
      >
        <p
          class="tm-li-tx__content__caption__title"
        >
          <div>
            
      Deposited 
      
            <b>
              100.00
            </b>
             
            <span>
               stakes
            </span>
          </div>
        </p>
      </div>
       
      <div
        class="tm-li-tx__content__information"
      >
        <div>
          
      On 
      
          <router-link
            to="/proposals/1"
          >
            
        Proposal #1
      
          </router-link>
        </div>
      </div>
    </div>
     
    <div
      class="tm-li-tx__content__block"
    >
      <a
        href="https://explorecosmos.network/blocks/56675"
      >
        
        Block #56675 
      
      </a>
      @ 00:00
    
    </div>
  </div>
</div>
`;

exports[`TmLiGovTransaction proposals should show proposals submission transactions 1`] = `
<div
  class="tm-li-tx"
>
  <div
    class="tm-li-tx__icon"
  >
    <img
      src="~assets/images/cosmos-logo.png"
      style="border-color: #3cb371;"
    />
  </div>
   
  <div
    class="tm-li-tx__content"
  >
    <div
      class="tm-li-tx__content__left"
    >
      <div
        class="tm-li-tx__content__caption"
      >
        <p
          class="tm-li-tx__content__caption__title"
        >
          <div>
            
      Submitted text proposal initial
      deposit 
            <b>
              100.00
            </b>
            <span>
               stakes
            </span>
          </div>
        </p>
      </div>
       
      <div
        class="tm-li-tx__content__information"
      >
        <div>
          
      Title: 
          <i>
            Test Proposal
          </i>
        </div>
      </div>
    </div>
     
    <div
      class="tm-li-tx__content__block"
    >
      <a
        href="https://explorecosmos.network/blocks/56673"
      >
        
        Block #56673 
      
      </a>
      @ 00:00
    
    </div>
  </div>
</div>
`;
